File: ui/candidate-portal/src/app/services/candidate-job-experience.service.ts
/*
 * Copyright (c) 2024 Talent Catalog.
 *
 * This program is free software: you can redistribute it and/or modify it under
 * the terms of the GNU Affero General Public License as published by the Free
 * Software Foundation, either version 3 of the License, or any later version.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE. See the GNU Affero General Public License
 * for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program. If not, see https://www.gnu.org/licenses/.
 */

import {Injectable} from '@angular/core';
import {environment} from "../../environments/environment";
import {HttpClient} from "@angular/common/http";
import {Observable} from "rxjs";
import {CandidateJobExperience} from "../model/candidate-job-experience";

@Injectable({
  providedIn: 'root'
})
export class CandidateJobExperienceService {

  private apiUrl: string = environment.apiUrl + '/job-experience';

  constructor(private http: HttpClient) { }

  createJobExperience(request): Observable<CandidateJobExperience> {
    return this.http.post<CandidateJobExperience>(`${this.apiUrl}`, request);
  }

  updateJobExperience(request: any): Observable<CandidateJobExperience> {
    return this.http.post<CandidateJobExperience>(`${this.apiUrl}/update`, request);
  }

  deleteJobExperience(id: number): Observable<any> {
    return this.http.delete<any>(`${this.apiUrl}/${id}`);
  }
}

File: server/src/main/java/org/tctalent/server/service/db/CandidateJobExperienceService.java
/*
 * Copyright (c) 2026 Talent Catalog.
 *
 * This program is free software: you can redistribute it and/or modify it under
 * the terms of the GNU General Public License as published by the Free
 * Software Foundation, either version 3 of the License, or any later version.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License
 * for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see https://www.gnu.org/licenses/.
 */

package org.tctalent.server.service.db;

import java.util.List;
import org.springframework.data.domain.Page;
import org.tctalent.server.model.db.CandidateJobExperience;
import org.tctalent.server.request.work.experience.CreateJobExperienceRequest;
import org.tctalent.server.request.work.experience.SearchJobExperienceRequest;
import org.tctalent.server.request.work.experience.UpdateJobExperienceRequest;

public interface CandidateJobExperienceService {

    Page<CandidateJobExperience> searchCandidateJobExperience(SearchJobExperienceRequest request);

    /**
     * Computes the context string for a given job experience.
     * <p>
     * The context string is used for generating embeddings or other processing.
     * @param experience Job experience
     * @return Context string.
     */
    String computeExperienceContext(CandidateJobExperience experience);

    CandidateJobExperience createCandidateJobExperience(CreateJobExperienceRequest request);

    CandidateJobExperience updateCandidateJobExperience(UpdateJobExperienceRequest request);

    CandidateJobExperience updateCandidateJobExperience(Long candidateId, UpdateJobExperienceRequest request);

    /**
     * Updates the embeddings for a list of candidate job experiences.
     * <p>
     * This method is intended for batch processing of candidate job experiences to update their
     * embeddings.
     * @param experiences the list of candidate job experiences to update. If empty, that signals
     *                    that the batch processing is complete and the currently building model
     *                    should be set to the status READY.
     */
    void batchUpdateCandidateJobExperienceEmbeddings(List<CandidateJobExperience> experiences);

    void deleteCandidateJobExperience(Long id);
}
